### ISSUE TITLE
[mono] Add support for library mode in the aot compiler

### ISSUE THREAD

**Issue by steveisok** [dotnet/runtime#79377]
As a native developer targeting Android, iOS, or wasm, I would like to use a shared library containing custom managed code plus the mono runtime so that I can call into managed functions from my native application. The library should take the same approach as NativeAOT, only exporting the symbols from managed functions decorated with `[UnmanagedCallersOnly(EntryPoint = "")]`. 

- [X] Run basic functional test loading two libraries in the same app (iOS & Android)
- [ ] Modify aot compiler to generate exports file with symbols for all `[UnmanagedCallersOnly(EntryPoint = "")]` 
- [ ] Create a library builder msbuild task that produces a cmake file to build / link the library
- [ ] Auto initialize and start the runtime when function is called from native code
- [ ] Extend direct-pinvoke support by accepting an list of functions & what static libraries to link
- [ ] Add workload to make it easier to install the mobile runtime packs + tooling
- [ ] Add functional test for library mode

**Comment by DeafMan1983** [comment]
Please do not forget to add includes and libraries if you use custom directories with includes and libraries for all platforms
Example: 
`ProjectRoot` means your created class library's project
```
<IncludePath directory="{ProjectRoot}/include" />
<LibraryPath directory="{ProjectRoot}/libs" />
```
Like for custom object c / headers for adapters like USB Device supports or any device supporting.... or plugins with platform-engine desktop/shell managements or for WebAssembly with functions for current browser's runtime like Electron or Chromium etc...

PS: Please put support with AppImage for Linux if you don't worry to get warning like not found library for custom shared libraries.
```
<IsLinux>
 <SingleAppImage>true</SingleAppImage>
</IsLinux>
```
I hope you have to add more features in Net 8.0

**Comment by davetao** [comment]
Hello @steveisok 

We are currently using the Embeddinator 4000 toolkit to make a shared library written in c# available as a native mobile SDK. Seeing the similarities in the use case, we were wondering if this was a suitable replacement seeing the E4K tool has been EOL for some time now.

**Comment by steveisok** [comment]
> Hello @steveisok
> 
> We are currently using the Embeddinator 4000 toolkit to make a shared library written in c# available as a native mobile SDK. Seeing the similarities in the use case, we were wondering if this was a suitable replacement seeing the E4K tool has been EOL for some time now.

It's a suitable replacement if all you want is a native shared library you can call into that doesn't rely on any android or ios bindings.  That's at least where we are starting from right now.

**Comment by curia-damiano** [comment]
Hi @steveisok , thank you for your answer and work on this issue.
I would have just a couple of questions:
- this would allow, for example, the porting of the .NET Azure SDK (previous creation of public methods) as the existing Azure SDK for Swift has been deprecated. As the .NET SDK is pure .NET, it doesn't contain any bindings, so it should work. Do you confirm?
- is it possible to test now the export to native libraries?
- do you plan to complete this work with the release of .NET 8?
- final question: what is the issue in case there are iOS or Android bindings?

**Comment by steveisok** [comment]
> * this would allow, for example, the porting of the .NET Azure SDK (previous creation of public methods) as the existing Azure SDK for Swift has been deprecated. As the .NET SDK is pure .NET, it doesn't contain any bindings, so it should work. Do you confirm?

Yes, as long as it doesn't contain bindings, then you should be able to expose entry points that in this case wrap the SDK.  

> * is it possible to test now the export to native libraries?

Yes, we have an experimental workload you can use.  There is a sample repo to help you get started https://github.com/steveisok/library-mode-sample

> * do you plan to complete this work with the release of .NET 8?

Yes, we are planning on having this feature for .NET 8. As of right now we are targeting iOS/MacCatalyst, Android, and Wasm.

> * final question: what is the issue in case there are iOS or Android bindings?

Try not to use them :-).  I'm not sure what the behavior will be.

**Comment by vitek-karas** [comment]
@steveisok do we have some kind of docs for this? Or is this issue the doc :-)

**Comment by steveisok** [comment]
We need a how to for mobile. Both for nativeaot and mono runtimes.

**Comment by ivanpovazan** [comment]
@steveisok @vitek-karas we have a tracking issue for it: https://github.com/dotnet/runtime/issues/96742
